File: C#/UnsupervisedCompetativeNeuralNet.Tests/UnsupervisedCompetativeNeuralNetTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using FluentAssert;
using NUnit.Framework;

namespace UnsupervisedCompetativeNeuralNet.Tests
{
    [TestFixture]
    public class UnsupervisedCompetativeNeuralNetTests
    {
        private const int MaxTries = 100;
        private const float LearningRate = 0.2f;
        private const float BiasLearningRate = 0.01f;

        [Test]
        public void ToyDataSet()
        {
            var dataset = new List<float[]>
            {
                new [] {1.0f, 0.9f, 0.8f},
                new [] {0.0f, 0.1f, 0.2f},
                new [] {0.2f, 0.1f, 0.1f},
                new [] {0.9f, 0.9f, 0.8f}
            };
            var expectedCluster = new List<int[]>
            {
                new [] {0, 3},
                new [] {1, 2}
            };

            TrainOnDataSet(dataset, expectedCluster).ShouldBeTrue();
        }

        [Test]
        public void OnlyASingleValueMatters()
        {
            var dataset = new List<float[]>
            {
                new [] {1.0f, 0.9f, 1.0f},
                new [] {1.0f, 1.0f, 0.0f},
                new [] {1.0f, 0.9f, 0.0f},
                new [] {0.9f, 1.0f, 1.0f}
            };
            var expectedCluster = new List<int[]>
            {
                new [] {0, 3},
                new [] {1, 2}
            };

            TrainOnDataSet(dataset, expectedCluster).ShouldBeTrue();
        }

        [Test]
        public void ThreeClasses()
        {
            var dataset = new List<float[]>
            {
                new[] {0.2f, 0.9f},
                new[] {0.2f, 0.8f},
                new[] {0.9f, 0.9f},
                new[] {0.8f, 0.9f},
                new[] {0.5f, 0.1f},
                new[] {0.3f, 0.0f}
            };
            var expectedCluster = new List<int[]>
            {
                new[] {0, 1},
                new[] {2, 3},
                new[] {4, 5}
            };

            TrainOnDataSet(dataset, expectedCluster).ShouldBeTrue();
        }

        private static bool TrainOnDataSet(IList<float[]> dataSet, IList<int[]> expectedClusters)
        {
            var unsupervisedNueralNet = new UnsupervisedCompetativeNeuralNet(dataSet.First().Length,
                expectedClusters.Count(), LearningRate, BiasLearningRate);
            for (var i = 0; i < MaxTries; i++)
            {
                unsupervisedNueralNet.Train(dataSet);

                if (ValidateClusters(dataSet, expectedClusters, unsupervisedNueralNet))
                    return true;
            }

            foreach (var item in dataSet)
            {
                var cluster = unsupervisedNueralNet.GetCluster(item);
                Console.WriteLine("{0} = {1}", item, cluster);
            }

            return false;
        }

        private static bool ValidateClusters(IList<float[]> dataSet, IEnumerable<int[]> expectedClusters, UnsupervisedCompetativeNeuralNet neuralNet)
        {
            return expectedClusters
                .All(expectedCluster =>
                        expectedCluster.Select(x => neuralNet.GetCluster(dataSet[x])).GroupBy(x => x).Count() == 1);
        }
    }
}

File: F#/UnsupervisedCompetativeNeuralNet.Tests/UnsupervisedCompetativeNeuralNetTests.fs
namespace UnsupervisedCompetativeNeuralNet.Tests
open UnsupervisedCompetativeNeuralNet
open NUnit.Framework
open System.Linq

[<TestFixture>]
type UnsupervisedCompetativeNeuralNetTests() = 
    let validate_clusters (data_set : float[][]) (expected_clusters : int[][]) (parameters : Parameters) =
        expected_clusters |> Array.

    let train_on_data_set (data_set : float[][]) (expected_clusters : int[][]) =
        let parameters = Parameters(data_set.[0].Length, expected_clusters.Length, 0.2, 0.01)
        for i in 0 .. 100 do
            train_many parameters data_set

            if 

    [<Test>]
    member this.ToyDataSet() = 
        let data_set = [| [|1.0; 0.9; 0.8|];
                          [|0.0; 0.1; 0.2|];
                          [|0.2; 0.1; 0.1|];
                          [|0.9; 0.9; 0.8|];
                        |]
        let expected_clusters = [| [|0; 3|];
                                   [|1; 2|]
                               |]
        Assert.IsTrue(train_on_data_set data_set expected_clusters)